checker/src/org/checkerframework/checker/lock/wholefiles/LockStore.java

```diff
@@ -0,0 +1,151 @@
+package org.checkerframework.checker.lock;
+
+/*>>>
+import org.checkerframework.checker.nullness.qual.Nullable;
+*/
+
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.ExecutableElement;
+
+import org.checkerframework.checker.lock.LockAnnotatedTypeFactory.SideEffectAnnotation;
+import org.checkerframework.checker.lock.qual.LockHeld;
+import org.checkerframework.dataflow.analysis.FlowExpressions;
+import org.checkerframework.dataflow.analysis.FlowExpressions.ArrayAccess;
+import org.checkerframework.framework.flow.CFAbstractStore;
+import org.checkerframework.framework.flow.CFValue;
+import org.checkerframework.framework.type.AnnotatedTypeFactory;
+import org.checkerframework.javacutil.AnnotationUtils;
+
+/*
+ * The Lock Store behaves like CFAbstractStore but requires the ability
+ * to insert exact annotations. This is because we want to be able to
+ * insert @LockPossiblyHeld to replace @LockHeld, which normally is
+ * not possible in CFAbstractStore since @LockHeld is more specific.
+ */
+public class LockStore extends CFAbstractStore<CFValue, LockStore> {
+
+    protected boolean inConstructorOrInitializer = false;
+
+    protected final AnnotationMirror LOCKHELD = AnnotationUtils.fromClass(analysis.getTypeFactory().getElementUtils(), LockHeld.class);
+
+    public LockStore(LockAnalysis analysis, boolean sequentialSemantics) {
+        super(analysis, sequentialSemantics);
+    }
+
+    /** Copy constructor. */
+    public LockStore(LockAnalysis analysis,
+            CFAbstractStore<CFValue, LockStore> other) {
+        super(other);
+        inConstructorOrInitializer = ((LockStore)other).inConstructorOrInitializer;
+    }
+
+    @Override
+    public LockStore leastUpperBound(LockStore other) {
+        LockStore newStore = super.leastUpperBound(other);
+
+        // Least upper bound of a boolean
+        newStore.inConstructorOrInitializer = this.inConstructorOrInitializer && other.inConstructorOrInitializer;
+
+        return newStore;
+    }
+
+    /*
+     * Insert an annotation exactly, without regard to whether an annotation was already present.
+     */
+    public void insertExactValue(FlowExpressions.Receiver r, AnnotationMirror a) {
```

Comment:
I would merge the other insertExactValue method with this on since this is the only call to the other.